prime-router/docs/design/validation/design.md

```diff
@@ -0,0 +1,218 @@
+   be routed to HHS Protect.
+   Resources:
+   Data Hub descriptions and Test Results Workflow Diagram
+</details>
+
+#### The HAPI HL7 library
+The HAPI library offers 3 "levels" of message validation:
+1. Basic Message Validation - very basic validation on message length and some segments, this is built in and will not 
+provide enough validation to meet our needs.
+2. Custom Message Validation - Extends basic and enabled through the extensibility offered by the message validation 
+framework itself. Allows you to override default behaviour.
+3. Validation by the use of Conformance Profiles - Unfortunately this boils down to essentially requiring 
+MicroSoft Workbench (for download: https://sourceforge.net/projects/mwbench/) which since we use Macs and a Google 
+search does not yield any alternatives, it seems like it rules this out as an option. If we can obtain access to a 
+Windows machine, it may be worth it. This tool appears to essentially be a plug and play tool that allows you to 
+specify a format and then will generate XML that can be used as a means to validate a message. Basically, it looks like 
+it provides a GUI to creating the specification.
+
+The HAPI HL7 library provides flexibility regarding the following:
+- Whether the message validation should be activated or not (it is turned on by default)
+- How to deal with unrecognized message versions, segments, fields, etc. during the parsing process
+- Behavior around how the parser should encode segments and fields even if no data content is present
+
+The HAPI HL7 library provides the following classes to validate HL7 messages:
+- `ValidationContext` - manages a collection of codified rules for handling messages, fields, and primitive types
+- `ValidationRuleBuilder` - what you use to create the validation rules
+- `MessageVisitor` - the interface the classes implement when they are in charge of doing the actual comparison
+- `ValidationException` - thrown when validation fails
+- `ValidationExceptionHandlerFactory` - interface that you can implement in order to control the exception handling
+
+Code Examples: <br/>
+A code example for custom message validation: 
+https://hapifhir.github.io/hapi-hl7v2/xref/ca/uhn/hl7v2/examples/CustomMessageValidation.html <br/>
+If you want to start with something more basic for a POC, here is an example that uses a default validation context
+rather than building one from scratch: https://hapifhir.github.io/hapi-hl7v2/xref/ca/uhn/hl7v2/examples/MessageValidation.html <br/>
+Here is an example of using the ValidationRuleBuilder: 
+https://hapifhir.github.io/hapi-hl7v2/xref/ca/uhn/hl7v2/examples/CustomMessageValidationWithValidatingVisitor.html <br/>
+An example for using conformance: 
+https://hapifhir.github.io/hapi-hl7v2/xref/ca/uhn/hl7v2/examples/MessageValidationUsingConformanceProfile.html <br/>
+
+#### NIST Validator
+This is a web tool that can be used for manual validation. However, as of the writing of this document, it is currently
```

Comment:
It sounds like DEX has this working? Ryan Harrison is a good contact for this but prefer looking at the DEX repo first for NIST implementation. I think the broken part is the RadxMars integration, which should be working in 2 weeks?

Reply:
Added more info to the doc on this.